lib/src/backends/record_replay/common.dart

```diff
@@ -83,3 +87,40 @@ class TypeMatcher<T> {
   /// Returns `true` if the given object is of type `T`.
   bool matches(dynamic object) => object is T;
 }
+
+/// Tells whether two objects are equal using deep equality checking.
+///
+/// Two lists are deeply equal if they have the same type, the same length, and
+/// every element in list A is pairwise deeply equal with the corresponding
+/// element in list B.
+///
+/// Two maps are deeply equal if they have the same type, the same length, the
+/// same set of keys, and the value for every key in map A is deeply equal to
+/// the corresponding value in map B.
+///
+/// All other types of objects are deeply equal if they are logically equal.
```

Comment:
I would be pedantic and make clear that they have to be of exactly the same runtime type, not just is-a.